src/System.Private.CoreLib/shared/System/Diagnostics/Tracing/CounterGroup.cs

```diff
@@ -205,48 +235,61 @@ private void OnTimer(object? state)
+                    } while(_nextPollingTimeStamp <= now);
                 }
             }
         }
 
-        #region PCL timer hack
 
-#if ES_BUILD_PCL
-        internal delegate void TimerCallback(object state);
 
-        internal sealed class Timer : CancellationTokenSource, IDisposable
-        {
-            private int _period;
-            private TimerCallback _callback;
-            private object _state;
+        private static Thread? s_pollingThread;
+        private static ManualResetEvent? s_pollingThreadEvent;
+        private static List<CounterGroup>? s_counterGroupEnabledList;
 
-            internal Timer(TimerCallback callback, object state, int dueTime, int period)
-            {
-                _callback = callback;
-                _state = state;
-                _period = period;
-                Schedule(dueTime);
-            }
-
-            private void Schedule(int dueTime)
+        private static void PollForValues()
+        {
+            ManualResetEvent? pollingThreadEvent = null;
+            while (true)
             {
-                Task.Delay(dueTime, Token).ContinueWith(OnTimer, null, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
-            }
+                int sleepDurationInMilliseconds = Int32.MaxValue;
+                lock (s_counterGroupLock)
+                {
+                    pollingThreadEvent = s_pollingThreadEvent;
 
-            private void OnTimer(Task t, object s)
-            {
-                Schedule(_period);
-                _callback(_state);
+                    if (s_counterGroupEnabledList != null)
+                    {
+                        foreach (CounterGroup counterGroup in s_counterGroupEnabledList)
+                        {
+                            DateTime now = DateTime.UtcNow;
+                            if (counterGroup._nextPollingTimeStamp < now + new TimeSpan(0, 0, 0, 0, 1))
+                            {
+                                counterGroup.OnTimer();
+                            }
+
+                            int millisecondsTillNextPoll = (int)((counterGroup._nextPollingTimeStamp - now).TotalMilliseconds);
+                            millisecondsTillNextPoll = Math.Max(1, millisecondsTillNextPoll);
+                            if (millisecondsTillNextPoll < sleepDurationInMilliseconds)
+                            {
+                                sleepDurationInMilliseconds = millisecondsTillNextPoll;
+                            }
+                        }
+                    }
+                }
+                // Only sleep if we actually had something in the enabled counter group
+                if (sleepDurationInMilliseconds != Int32.MaxValue)
+                {
+                    Thread.Sleep(sleepDurationInMilliseconds);
```

Comment:
Sorry, the pattern continues, I realized another issue. If one session asks for a slow rate of counters (say once an hour), and then a 2nd session asks for a faster rate (say once a second), this sleep has no way to be interrupted. You probably need to wait on an AutoResetEvent using the sleepDuration as a timeout so that you can wake up after this period of time OR when the event is signaled for a polling interval change.